cmd/tetra/debug/maps.go

```diff
@@ -0,0 +1,235 @@
+
+	"github.com/cilium/tetragon/pkg/bugtool"
+	"github.com/spf13/cobra"
+	"golang.org/x/exp/maps"
+)
+
+const tetragonBPFFS = "/sys/fs/bpf/tetragon"
+
+func NewMapCmd() *cobra.Command {
+	var lines int
+
+	cmd := cobra.Command{
+		Use:     "maps",
+		Aliases: []string{"map"},
+		Short:   "Retrieve information about maps on the host related to tetragon",
+		Long: fmt.Sprintf(`Retrieve information about maps on the host related to tetragon.
+
+- AllMaps: all the BPF maps loaded on the host
+- PinnedProgsMaps: all the BPF maps used in programs pinned under %[1]s
+- PinnedMaps: all the BPF maps pinned under %[1]s
+
+The first table is the total of memlock bytes for the maps, those values are
+accurate mostly from v6.4[^1].
+
+The second table is in number of maps, inter, exter and union are the respective
+math operations on the two first set, diff is the number of maps that were
+detected as pinned maps but not used in pinned progs, ideally this number should
+be 0. Note that the union should be the set of all the tetragon related maps on
+the host if all the programs are properly pinned under the bpffs and all the
+unused maps are also pinned under the bpffs.
+
+The third table is the details of the maps that are in the diff, which means
+maps that are pinned but unreferenced from pinned programs. Ideally, this table
+should be empty.
+
+The fourth table aggregate map by name and add the count column indicating how
+many maps were aggregated for the given name. It ranks map from the one
+consuming the most memory to the one consuming the less. Use the -n flag to
+adjust the number of item in the table.
+
+[^1]: https://lore.kernel.org/all/20230305124615.12358-1-laoar.shao@gmail.com/`, tetragonBPFFS),
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			// check that the bpffs exists and we have permissions
+			_, err := os.Stat(tetragonBPFFS)
+			if err != nil {
+				return fmt.Errorf("make sure tetragon is running and you have enough permissions: %w", err)
+			}
+
+			// print BPF maps memory usage
+			allMapsMem, err := bugtool.TotalByteMemlockAllMaps()
+			if err != nil {
+				return fmt.Errorf("failed to retrieve all maps memlock: %w", err)
+			}
+			pinnedProgsMapsMem, err := bugtool.TotalByteMemlockMapsFromProgs(tetragonBPFFS)
+			if err != nil {
+				return fmt.Errorf("failed to retrieve maps from pinned prog memlock: %w", err)
+			}
+			pinnedMapsMem, err := bugtool.TotalByteMemlockMaps(tetragonBPFFS)
+			if err != nil {
+				return fmt.Errorf("failed to retrieve pinned maps memlock: %w", err)
+			}
+			w := tabwriter.NewWriter(cmd.OutOrStdout(), 0, 0, 3, ' ', 0)
+			fmt.Fprintln(w, "AllMaps\tPinnedProgsMaps\tPinnedMaps")
+			fmt.Fprintf(w, "%d\t%d\t%d\n", allMapsMem, pinnedProgsMapsMem, pinnedMapsMem)
+			w.Flush()
+			cmd.Println()
+
+			// print details on map distribution
+			pinnedProgsMapsSet := map[int]bugtool.ExtendedMapInfo{}
+			pinedProgsMaps, err := bugtool.FindMapsUsedByPinnedProgs(tetragonBPFFS)
+			if err != nil {
```

Comment:
we already iterated that as part of TotalByteMemlockMapsFromProgs right? same for FindPinnedMaps and TotalByteMemlockMaps? could we perhaps run that just once at beginning and use the []ExtendedMapInfo for both memlock sum and printing later?

Reply:
yep indeed the `Total`s are pretty much wrappers around the other function + sum. It could be optimized and only done once at the beginning and the sum computed locally with the results. Maybe I should remove the `Total` functions

Reply:
fixed it by removing them